app/models/facility.rb

```diff
@@ -179,22 +179,93 @@ def teleconsultation_enabled?
   end
 
   def teleconsultation_phone_number_with_isd
-    return if teleconsultation_isd_code.blank? || teleconsultation_phone_number.blank?
+    teleconsultation_phone_number = teleconsultation_phone_numbers.first
+    return if teleconsultation_phone_number.blank?
+
+    Phonelib.parse(teleconsultation_phone_number.isd_code + teleconsultation_phone_number.phone_number).full_e164
+  end
 
-    Phonelib.parse(teleconsultation_isd_code + teleconsultation_phone_number).full_e164
+  def teleconsultation_phone_numbers_with_isd
+    teleconsultation_phone_numbers.map do |phone_number|
+      {phone_number: Phonelib.parse(phone_number.isd_code + phone_number.phone_number).full_e164}
+    end
   end
 
   CSV::Converters[:strip_whitespace] = ->(value) {
     begin
-                value.strip
+      value.strip
     rescue
       value
-              end
+    end
   }
 
+  def teleconsultation_phone_numbers
+    read_attribute(:teleconsultation_phone_numbers).map { |t| TeleconsultationPhoneNumber.new(t) }
+  end
+
+  def teleconsultation_phone_numbers_attributes=(attributes)
+    teleconsultation_phone_numbers = []
+    attributes.each do |_index, attrs|
+      next if attrs.delete("_destroy") == "true"
+
+      teleconsultation_phone_numbers << TeleconsultationPhoneNumber.new(attrs)
+    end
+
+    write_attribute(:teleconsultation_phone_numbers, teleconsultation_phone_numbers)
+  end
+
+  def build_teleconsultation_phone_number
+    numbers = teleconsultation_phone_numbers.dup
+    numbers << TeleconsultationPhoneNumber.new({isd_code: Rails.application.config.country["sms_country_code"],
+                                                phone_number: ""})
+    self.teleconsultation_phone_numbers = numbers
+  end
+
+  class TeleconsultationPhoneNumber
```

Comment:
This class and its uses are now _just_ for feeding _into_ form elements on the page using `form_for`, right? Do you think we can remove all the custom attribute reading and writing, and add a single method to the facility class, or controller, that's like `TeleconsultationPhoneNumbersForForm` or something? Overall, I'm wondering if we can avoid so much model behavior to power one form in the view layer.

Reply:
I've refactored this class to be a struct instead, but I haven't been able to figure out a way to use this without the custom accessors.

The only other way I can think of is to do some client-side form hacking to make this a single field and serialize and deserialize its contents as needed. This will remove the need for the accessors in Facility, and move what they're doing to a separate serializer class.

This will make the form more complex though, since it will have to parse and assemble json for the single field, and we'd still want something like the struct to keep the field clean.